src/v/cluster/controller.cc

```diff
@@ -454,39 +457,111 @@ ss::future<> controller::stop() {
-      || _raft0->config().brokers().size() > 1) {
-        // The controller log has already been written to
+    }
+
+    if (config::node().empty_seed_starts_cluster()) {
+        // This is the legacy empty-seed-starts-cluster mode
+
+        if (!config::node().seed_servers().empty()) {
+            // We are not on the root node
+            co_return;
+        }
+        if (
+          _raft0->last_visible_index() > model::offset{}
+          || _raft0->config().brokers().size() > 1) {
+            // The controller log has already been written to
+            co_return;
+        }
+
+        // Internal RPC does not start until after controller startup
+        // is complete (we are called during controller startup), so
+        // it is guaranteed that if we were single node/empty controller
+        // log at start of this function, we will still be in that state
+        // here.  The wait for leadership is really just a wait for the
+        // consensus object to finish writing its last_voted_for from
+        // its self-vote.
+        while (!_raft0->is_leader()) {
+            co_await ss::sleep(100ms);
+        }
+
+        auto err
+          = co_await _security_frontend.local().maybe_create_bootstrap_user();
+        vassert(
+          err == errc::success,
+          "Controller write should always succeed in single replica state "
+          "during creation");
+
+        co_return co_await create_cluster();
+    }
+
+    if (!local_node_is_seed_server) {
         co_return;
     }
 
-    // Internal RPC does not start until after controller startup
-    // is complete (we are called during controller startup), so
-    // it is guaranteed that if we were single node/empty controller
-    // log at start of this function, we will still be in that state
-    // here.  The wait for leadership is really just a wait for the
-    // consensus object to finish writing its last_voted_for from
-    // its self-vote.
-    while (!_raft0->is_leader()) {
+    // The new cluster creation is done by the leader of raft0 consisting of
+    // seed servers, once elected
+    while (!_bootstrap_backend.local().get_cluster_uuid()
+           && !_raft0->is_leader()) {
         co_await ss::sleep(100ms);
     }
 
-    auto err
-      = co_await _security_frontend.local().maybe_create_bootstrap_user();
-    vassert(
-      err == errc::success,
-      "Controller write should always succeed in single replica state during "
-      "creation");
+    if (_bootstrap_backend.local().get_cluster_uuid()) {
+        vlog(
+          clusterlog.info,
+          "Cluster is {}",
+          *_bootstrap_backend.local().get_cluster_uuid());
+        co_return;
+    }
+
+    if (
+      _raft0->last_visible_index() > model::offset{}
+      || _raft0->config().brokers().size() > 1) {
+        // The controller log has already been written to
+        // TODO: create_cluster() but use the metrics' cluster_id
```

Comment:
We have discussed before that it's convenient to have the identical values for cluster ID and cluster UUID, and for that purpose the latter should be initialized from the former in the cluster upgrade scenario.

Reply:
I think we're still not really on the same page here yet. I was under the impression we would never initialize a `cluster_uuid` with anything other than a random value. Reconciliation of `cluster_id` and `cluster_uuid` would take the form of: in metrics only, if we have a `cluster_id`, use it, if not, use the `cluster_uuid`, where it is possible that `cluster_id` and `cluster_uuid` are not identical.

We have no guarantee that `cluster_id` is even a UUID, since it can be set by the user.